src/compiler/checker.ts

```diff
@@ -7039,12 +7041,28 @@ namespace ts {
                 hasSpreadElement = hasSpreadElement || e.kind === SyntaxKind.SpreadElementExpression;
             }
             if (!hasSpreadElement) {
+                // If array literal is actually a destructuring pattern, mark it as an implied type. We do this such
+                // that we get the same behavior for "var [x, y] = []" and "[x, y] = []".
+                if (inDestructuringPattern && elementTypes.length) {
+                    return createImpliedType(createTupleType(elementTypes));
+                }
                 let contextualType = getContextualType(node);
-                if (contextualType && contextualTypeIsTupleLikeType(contextualType) || inDestructuringPattern) {
-                    return createTupleType(elementTypes);
+                let contextualTupleLikeType = contextualType && contextualTypeIsTupleLikeType(contextualType) ? contextualType : undefined;
+                if (contextualTupleLikeType) {

```

Comment:
I think it's simpler to just do
```ts
if (contextualType && contextualTypeIsTupleLikeType(contextualType)) {

}
```
And just refer to contextualType inside. Essentially just inline the condition